modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/near/GridNearTxFinishFuture.java

```diff
@@ -978,6 +987,36 @@ public GridDistributedTxMapping mapping() {
 
         /** {@inheritDoc} */
         @Override boolean onNodeLeft(UUID nodeId, boolean discoThread) {
+            if (tx.state() == COMMITTING || tx.state() == COMMITTED) {
+                Map<UUID, Collection<UUID>> txNodes = tx.transactionNodes();
+
+                Collection<UUID> backups = txNodes.get(nodeId);
+
+                boolean hasBackups = false;
+
+                if (backups != null)
+                    hasBackups = backups.stream().anyMatch(backupId -> cctx.discovery().node(backupId) != null);
+
+                if (cctx.discovery().node(m.primary().id()) == null && !hasBackups) {
+                    String strTxEntry = "";
+
+                    Iterator<IgniteTxEntry> entryIter = m.entries().iterator();
+
+                    if (entryIter.hasNext()) {
+                        IgniteTxEntry firstTxEntry = entryIter.next();
+
+                        strTxEntry = " [cacheName=" + firstTxEntry.cached().context().name() +
+                            ", partition=" + firstTxEntry.key().partition() +
+                            (S.includeSensitive() ? ", key=" + firstTxEntry.key() : "") +
+                            "]";
+                    }
+
+                    onDone(new CacheInvalidStateException(ALL_PARTITION_OWNERS_LEFT_GRID_MSG + strTxEntry));
+
+                    return true;
+                }
+            }
```

Comment:
Too much of lines.
This can be simplified to something like:

if (concat(of(m.primary().id()), m.backups().stream()).noneMatch(uuid -> cctx.discovery().alive(uuid))) {
                    onDone(new CacheInvalidStateException(ALL_PARTITION_OWNERS_LEFT_GRID_MSG +
                        m.entries().stream().map(e -> " [cacheName=" + e.cached().context().name() +
                            ", partition=" + e.key().partition() +
                            (S.includeSensitive() ? ", key=" + e.key() : "") +
                            "]").findFirst().orElse("")));

                    return true;
                }